core/raftlease/store.go

```diff
@@ -0,0 +1,261 @@
+		result[k] = v
+	}
+	return result
+}
+
+// Refresh is part of lease.Store.
+func (s *Store) Refresh() error {
+	return nil
+}
+
+// Advance is part of globalclock.Updater.
+func (s *Store) Advance(duration time.Duration) error {
+	s.prevTimeMu.Lock()
+	defer s.prevTimeMu.Unlock()
+	newTime := s.prevTime.Add(duration)
+	err := s.forwardCommand(&Command{
+		Version:   CommandVersion,
+		Operation: OperationSetTime,
+		OldTime:   s.prevTime,
+		NewTime:   newTime,
+	})
+	if err == globalclock.ErrConcurrentUpdate {
+		// Someone else updated before us - get the new time.
+		s.prevTime = s.fsm.GlobalTime()
+	} else if err == nil {
+		s.prevTime = newTime
+	}
+	return errors.Trace(err)
+}
+
+func (s *Store) forwardCommand(command *Command) error {
+	bytes, err := command.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	requestID := atomic.AddUint64(&s.requestID, 1)
+	responseTopic := s.config.ResponseTopic(requestID)
+
+	responseChan := make(chan ForwardResponse, 1)
+	errChan := make(chan error)
+	unsubscribe, err := s.hub.Subscribe(
+		responseTopic,
+		func(_ string, resp ForwardResponse, err error) {
+			if err != nil {
+				errChan <- err
+				return
+			}
+			responseChan <- resp
+		},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer unsubscribe()
+
+	_, err = s.hub.Publish(s.config.RequestTopic, ForwardRequest{
+		Command:       bytes,
+		ResponseTopic: responseTopic,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	select {
+	case <-s.config.Clock.After(s.config.ForwardTimeout):
+		return lease.ErrTimeout
+	case err := <-errChan:
+		return errors.Trace(err)
+	case response := <-responseChan:
+		return RecoverError(response.Error)
+	}
+}
+
+// ForwardRequest is a message sent over the hub to the raft forwarder
+// (only running on the raft leader node).
+type ForwardRequest struct {
+	Command       []byte `yaml:"command"`
+	ResponseTopic string `yaml:"response-topic"`
+}
+
+// ForwardResponse is the response sent back from the raft forwarder.
+type ForwardResponse struct {
+	Error *ResponseError `yaml:"error"`
+}
+
+// ResponseError is used for sending error values back to the lease
+// store via the hub.
+type ResponseError struct {
+	Message string `yaml:"message"`
+	Code    string `yaml:"code"`
+}
+
+// AsResponseError returns a *ResponseError that can be sent back over
+// the hub in response to a forwarded FSM command.
+func AsResponseError(err error) *ResponseError {
+	if err == nil {
+		return nil
+	}
+	message := err.Error()
+	var code string
+	switch errors.Cause(err) {
+	case lease.ErrInvalid:
+		code = "invalid"
+	case globalclock.ErrConcurrentUpdate:
+		code = "concurrent-update"
+	default:
+		code = "error"
+	}
+	return &ResponseError{
+		Message: message,
+		Code:    code,
+	}
+}
+
+// RecoverError converts a ResponseError back into the specific error
+// it represents, or into a generic error if it wasn't one of the
+// singleton errors handled.
+func RecoverError(resp *ResponseError) error {
```

Comment:
does this need to be exported?
(same to a bunch of other things in here)

Reply:
ForwardRequest, ForwardResponse, ResponseError and AsResponseError need to be exported so they can be used from the raftforwarder. RecoverError maybe doesn't? But it seems weird to export a conversion function in one direction and not the reverse. Not sure.